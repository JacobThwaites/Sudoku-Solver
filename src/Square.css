.square {
    max-width: 50px;
}