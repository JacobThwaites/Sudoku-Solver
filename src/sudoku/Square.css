.square {
  margin: 0;
  padding: 0;
  border-collapse: collapse;
  box-sizing: border-box;
  font-size: 3em;
  font-family: Helvetica;
  text-align: center;
  width: 11.11%;
  height: 11.201%;
  border: 1px solid #444;
  caret-color: transparent;
}

.square {
  font-weight: lighter;
}

.starting-position {
  font-weight: bold;
}

.top-square {
  border-top: 3px solid black;
}

.bottom-square {
  border-bottom: 3px solid black;
}

.left-square {
  border-left: 3px solid black;
}

.right-square {
  border-right: 3px solid black;
}

@media (max-width: 620px) {
  .square {
    width: 10.11vw;
    height: 10.201vw;
  }
}

@media (max-width: 400px) {
  .square {
    font-size: 2rem;
  }
}

@media (max-width: 300px) {
    .square {
      font-size: 1.5rem;
    }
  }
