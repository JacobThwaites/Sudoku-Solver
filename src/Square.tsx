import React from 'react';
import './Square.css'

interface SquareProps {
  value: number,
  index: number,
  handleChange: Function, 
}

function Square(props: SquareProps) {
  function handleInput(event: any) {
    const input = Number(event.target.value);

    if (!input || input > 9) {
      return;
    }

    props.handleChange(input, props.index);
  }

  return (
    <>
      <input 
        className='square'
        type="text" 
        value={props.value}
        maxLength={1} 
        onChange={handleInput}
      />
    </>
  );
}

export default Square;
