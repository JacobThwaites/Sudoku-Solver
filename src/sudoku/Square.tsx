import React, { useRef, useEffect } from 'react';
import { isArrowKeyPress, isBackspacePress, isNumberKeyPress } from '../utils/keypressValidation';
import './Square.css'


interface SquareProps {
  value: number,
  column: number,
  setSquareValue: Function, 
  handleArrowKey: Function,
  className: string,
  setActiveSquare: any,
  isActiveSquare: boolean
}

export default function Square(props: SquareProps) {
  const ref = useRef(null);

  useEffect(() => {
    if (props.isActiveSquare) {
      // Move element into view when it is focused
      ref.current.focus();
    }
  }, [props.isActiveSquare]);

  function handleKeyDown(event: any) {
    if (isArrowKeyPress(event.keyCode)) {
      props.handleArrowKey(event.keyCode);
      return;
    }

    if (isBackspacePress(event.keyCode)) {
      props.setSquareValue(null, props.column);
      return;
    }

    if (isNumberKeyPress(event.keyCode)) {
      props.setSquareValue(Number(event.key), props.column);
      return;
    }
  }

  return (
    <>
      <input 
        className={props.className}
        type="text" 
        value={props.value ? props.value : ''}
        maxLength={1} 
        ref={ref}
        // onChange handled by onKeydown
        onChange={() => {}}
        onKeyDown={handleKeyDown}
        onClick={props.setActiveSquare}
      />
    </>
  );
}