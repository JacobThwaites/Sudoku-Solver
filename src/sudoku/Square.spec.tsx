import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Square from './Square';


it('renders Square correctly', () => {
    const component = renderer.create(
        <Square value={1} column={1} setSquareValue={jest.fn()} handleArrowKey={jest.fn()} className="testClass" setActiveSquare={jest.fn()} isActiveSquare={false} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('handles number input', () => {
    const setSquareValue = jest.fn();
    const key = 3;
    const columnNumber = 2;

    const { getByTestId } = render(
        <Square value={1} column={columnNumber} setSquareValue={setSquareValue} handleArrowKey={jest.fn()} className="testClass" setActiveSquare={jest.fn()} isActiveSquare={false} />
    );

    const square = getByTestId('square');
    fireEvent.keyDown(square, { key: key, keyCode: 51 });
    expect(setSquareValue).toHaveBeenCalledWith(key, columnNumber);
});

it('handles backspace press', () => {
    const setSquareValue = jest.fn();
    const backSpaceCode = 8;
    const columnNumber = 2

    const { getByTestId } = render(
        <Square value={1} column={columnNumber} setSquareValue={setSquareValue} handleArrowKey={jest.fn()} className="testClass" setActiveSquare={jest.fn()} isActiveSquare={false} />
    );

    const square = getByTestId('square');
    fireEvent.keyDown(square, { key: backSpaceCode, keyCode: backSpaceCode });
    expect(setSquareValue).toHaveBeenCalledWith(null, columnNumber);
});

it('handles arrowkey press', () => {
    const handleArrowKey = jest.fn();
    const arrowUpKey = 37;

    const { getByTestId } = render(
        <Square value={1} column={2} setSquareValue={jest.fn()} handleArrowKey={handleArrowKey} className="testClass" setActiveSquare={jest.fn()} isActiveSquare={false} />
    );

    const square = getByTestId('square');
    fireEvent.keyDown(square, { key: arrowUpKey, keyCode: arrowUpKey });
    expect(handleArrowKey).toHaveBeenCalledWith(arrowUpKey);
});