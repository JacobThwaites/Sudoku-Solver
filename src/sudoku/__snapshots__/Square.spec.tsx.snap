// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Square correctly 1`] = `
<input
  className="testClass"
  data-testid="square"
  maxLength={1}
  onChange={[Function]}
  onClick={[MockFunction]}
  onKeyDown={[Function]}
  type="text"
  value={1}
/>
`;
